pkg/factory/config.go

```diff
@@ -477,9 +477,7 @@ func (u *UserPlaneInformation) validate() (bool, error) {
 // UPNode represent the user plane node
 type UPNode struct {
 	Type                 string                  `json:"type" yaml:"type" valid:"upNodeType,required"`
-	NodeID               string                  `json:"nodeID" yaml:"nodeID" valid:"host,optional"`
-	Addr                 string                  `json:"addr" yaml:"addr" valid:"host,optional"`
-	ANIP                 string                  `json:"anIP" yaml:"anIP" valid:"host,optional"`
+	NodeID               string                  `json:"nodeID" yaml:"nodeID" valid:"host"`
```

Comment:
We can't replace Addr & ANIP with NodeId.
NodeId may not be address for this UPNode.

Reply:
So what exactly was the original intention behind `NodeID`, `Addr`, and `ANIP`?

I thought that NodeID was the address that the 5GC uses to communicate with a node (gNB - N1 link, UPF - N4 link).

I agree that with only `NodeID`, there will only be a single address for a node, which does not work in the gNB case.

This part of the configuration was very confusing for me at the beginning, which is why I wanted to make it easier to understand for users. You basically define gNB and UPF in the same way, but `ANIP` only exists for gNBs and `Addr`, `InterfaceUpfInfoList`, `Dnn`, and `SNssaiInfos`only exists for UPFs.

Reply:
I think `NodeId` is ID to identify the node. 
- `Addr` is for UPF to set UPF's address (IPv4, IPv6, FQDN)
-  `ANIP` is for gNB.

So the major thing is that we can't use NodeId as Address, but we can keep `Addr` for use of UPF & gNB.

Reply:
I don't think the SMF needs to know the AN IP address (of the N3 interface) in the configuration at all. For the GTP-U tunnel, the information is set in `func (t *UPTunnel) UpdateANInformation(ip net.IP, teid uint32)`, and the `UPNode.ANIP` value that exists in the current code is never used, the variable is only read from the config file.

Reply:
I also noticed that the `ANIP` is never used. 
However, we can now have the `Addr` attribute as `ANIP` when the Node is gNB. 
As for the usage of `ANIP` in SMF, I can discuss it with the team and get back to you. 

So, here we should add `Addr` back and use it as the origin `Addr` or `ANIP.`

Reply:
What I don't understand: The `NodeID` struct is defined in the `pfcpType` package at `github.com/free5gc/pfcp/pfcpType`. It would make sense to have a NodeID for a UPF only, as it is the only node that uses PFCP. The gNB does not communicate over PFCP with the SMF or other NFs in the core network.

For me, it would make sense to have the following yaml configuration:
```yaml
userplaneInformation:
  upNodes:
    gNB1: # the name of the node (string)
      type: AN # the type of the node (AN or UPF)
    UPF: # the name of the node (string)
      type: UPF # the type of the node (AN or UPF)
      nodeID: 127.0.0.8 # the address of this UPF for PFCP communication
      sNssaiUpfInfos: 
        [...]
      interfaces: # user plane interfaces for this UPF
        [...]
  links: # the topology graph of userplane, A and B represent the two nodes of each link
    - A: gNB1
      B: UPF
```

The gNB UPNode configuration imho should not have an additional `anIP` value. The SMF actually does not communicate with the gNB, and there is no need to have this information in its config. Actually it would remove the effort for the free5gc users, as they do not need to adapt this value.

Reply:
I actually see that there is additional confusion for me:
```yaml
pfcp: # the IP address of N4 interface on this SMF (PFCP)
  nodeID: 127.0.0.1 # the Node ID of this SMF
  listenAddr: 127.0.0.1 # the IP/FQDN of N4 interface on this SMF (PFCP)
  externalAddr: 127.0.0.1 # the IP/FQDN of N4 interface on this SMF (PFCP)
```

What is the difference between these addresses? Can we reduce it to only the NodeID (in a separate PR maybe)?

Reply:
The main problem here is that `nodeID` can't used as the IP address. 
So the reasonable configuration to me is like:
```yaml
userplaneInformation:
  upNodes:
    gNB1: # the name of the node (string)
      type: AN # the type of the node (AN or UPF)
      nodeID: gnb1 # the node ID of gNB
    UPF: # the name of the node (string)
      type: UPF # the type of the node (AN or UPF)
      nodeID: 127.0.0.8 # the node ID of this UPF
      addr: 127.0.0.8 # the address of this UPF for PFCP communication
      sNssaiUpfInfos: 
        [...]
      interfaces: # user plane interfaces for this UPF
        [...]
  links: # the topology graph of userplane, A and B represent the two nodes of each link
    - A: gNB1
      B: UPF
```

Reply:
Is it ok if I just keep asking questions here, or should we rather take this offline and away from GitHub?

I don't understand why we can't use the NodeID as the PFCP address. The requirements for a NodeID are that it is a [valid host name](https://github.com/free5gc/smf/blob/main/pkg/factory/config.go#L480), so it must either be an IPv4 or IPv6 address or an FQDN. 
I see no need for having a second address parameter for the UPF. It can be identified through the PFCP address (the NodeID). And therefore, the gNB does not need this NodeID parameter, as it has no N4 interface.

Reply:
It's totally fine to talk about this here, don't worry. I have email and LinkedIn on my profile if you want. 

Sorry, I didn't notice that the `nodeId` has to be `host` in the validator. 
I think your design works to simplify the configuration; thanks for pointing it out. 

We should point out in the description that the configuration has some changes and remember to correct the sample configuration before the next release. (I will do this if the PR is merged).